File: Algorithms/src/3699. Number of ZigZag Arrays I/Code/solution.py
class Solution:
    def zigZagArrays(self, n: int, l: int, r: int) -> int:
        MOD = 1000000007
        m = r - l + 1

        # Length 1: every rank is valid
        dp = [1] * m

        for _ in range(2, n + 1):
            # Reverse so one prefix-sum pass handles
            # the alternating transition automatically
            dp.reverse()

            pref = 0

            for i in range(m):
                old = dp[i]

                # New state gets all previous contributions
                dp[i] = pref

                pref = (pref + old) % MOD

        # Sum all ending positions
        ans = sum(dp) % MOD

        # Count both zigzag directions
        return (ans * 2) % MOD

File: Algorithms/src/3699. Number of ZigZag Arrays I/Code/solution.js
/**
 * @param {number} n
 * @param {number} l
 * @param {number} r
 * @return {number}
 */
var zigZagArrays = function (n, l, r) {
  const MOD = 1000000007;
  const m = r - l + 1;

  // Length 1 DP
  const dp = Array(m).fill(1);

  for (let len = 2; len <= n; len++) {
    // Reverse to reuse the same prefix-sum transition
    dp.reverse();

    let pref = 0;

    for (let i = 0; i < m; i++) {
      const old = dp[i];

      // New state = accumulated prefix
      dp[i] = pref;

      pref = (pref + old) % MOD;
    }
  }

  let ans = 0;

  for (const x of dp) {
    ans = (ans + x) % MOD;
  }

  // Both possible starting directions
  return (ans * 2) % MOD;
};
